<a href="https://colab.research.google.com/github/temperedangel/Github-and-Jupyter-setup/blob/main/module2_titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Human-Centered Data Science



Welcome to our exploration of the Titanic dataset. We're going to walk through the data science lifecycle using this dataset as an example. With that in mind, the first thing we want to do is determine what question(s) we want to ask.  


## What's the Question?





Everyone knows that the Titanic sank after running into an iceberg. Many people lost their lives in the accident, but, spectacularly, many people were able to survive the accident.  

That seems like a good place to start, some simple **descriptive** questions to understand more about the event. For example, we can ask:
1. How many people were on the Titanic?
2. How many survived the accident?
3. How many fatalities were there?

We can also ask an **exploratory** question:
1. Did more men, women, or children die/survive in the accident?

And we could ask a **predictive** question: Can we predict whether a passenger would have lived or survived given what we know about them? We might want to know this if we're a safety inspector working for a modern cruise company. Can we learn something about the survival rates on the Titanic that can generalize to new ocean liners?

There are other possibilities, and you may have come up with different ones in your earlier reflections. For now, let's keep these and start learning how to answer them with data.




## Collect Data

Before we can work with data, we need to get our notebook ready by loading some essential libraries. If you're new to programing, you can think of a library as a toolbox of useful functions that you can borrow and use in your work. The beauty of libraries is that someone else has done the work of creating the tools: You get to use them without needing to build them yourself.

In this example, we'll load one of the most important libraries in data science, the Python Data Analysis Library, or `pandas`. You can guess from the name that `pandas` provides tools for data analysis and manipulation. `pandas` is a little long time to type repeatiedly, so you'll often see people import the library and then refer to it by the shorter `pd`.

In [1]:
import pandas as pd

Some of the many important tools in `pandas` allow us to read and write data files. We can read pdf, Excel, text...all sorts of files. For now, we want to read data from a comma-separated values, or CSV, file. `pandas` has a simple function named `read_csv` to read from a CSV file. We'll use it to read in a file of data related to the Titanic.

`pd.read_csv()` means "call the `read_csv()` function that sits in the `pd` (`pandas`) library. We're going to create a variable called `titanic_df` and set its value to our data table. We'll name the variable `titanic_df` to note that it points to a pandas dataframe (df) containing Titanic data.

In [2]:
# The giant URL below points to a data file named "titanic3.csv"
titanic_df_URL = "https://raw.githubusercontent.com/bsmith-classroom/datasets/main/titanic3.csv"

titanic_df = pd.read_csv(titanic_df_URL)

Now that we have the data, we should look to see what it looks like. `pandas` has a function named `head` that shows the first few rows of data.

In [3]:
print(titanic_df.head())

   pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St 

The `info` function displays useful information about the dataset. We can see the data types for each column, for example, the `age` data are currently `float64` numbers.   



In [4]:
print(titanic_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB
None


We can also see the number of **non-null** entries in each column. Null, or **missing**, values can affect our data analyses later. We'll try to avoid them for them moment. But you can imagine how the missing values complicate things later. As an example, look at the `age` column above: It only has 1046 non-null values out of a possible 1309. That means we don't have age information for every passenger. We'll return to this problem later. First, let's look at the meanings of each column.

### The Data Dictionary
We see there are fourteen columns with names `pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body, and home.dest`. We need to know a little more about each column name. That's what *data dictionaries* are for: They define the contents of each column. In this case, the data dictionary provides this information:

| Column Name   | Description                                               |
|---------------|-----------------------------------------------------------|
| pclass   | Passenger Class (1 = 1st; 2 = 2nd; 3 = 3rd)         
| survived | Did the passenger survive? (0 = No; 1 = Yes)  
| name  | Name  
sex | Passenger gender  
age | Age (in years; fractional if age less than one (1). If the age is estimated, it is in the form xx.5)  
sibsp | Number of Siblings/Spouses Aboard  
parch | Number of Parents/Children Aboard   
ticket | Ticket Number  
fare | Passenger Fare (pre-1970 British pound)  
cabin | Cabin number  
embarked | Port of Embarkation (C = Cherbourg; Q = Queenstown; S = Southampton)  
boat | Lifeboat ID if the passenger survived   
body | Body number if the passenger didn't survive and the body was recovered   
home.dest | the home of the passenger, occasionally followed by their ultimate destination


Earlier, we used `info` and saw the number of non-null values in each column. We can also see the number of null values using the pandas function `isnull`.

In [5]:
print(titanic_df.isnull().sum())

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


`isnull` returns `True` if a data value is missing/undefined. We modify if with `sum()` to get a count of all the missing values. We can see that there are 263 missing ages and 1014 missing cabin values.

In the dataset, this undefined value is represented by `NaN`, which stands for Not a Number. Take a look:

In [6]:
print(titanic_df.iloc[886])

pclass                         3
survived                       1
name         Jermyn, Miss. Annie
sex                       female
age                          NaN
sibsp                          0
parch                          0
ticket                     14313
fare                        7.75
cabin                        NaN
embarked                       Q
boat                           D
body                         NaN
home.dest                    NaN
Name: 886, dtype: object


The command `iloc` stands for "integer location." Calling `iloc[886]` will return the passenger at the 886th row of data. Ok, it's really the *887th* row because we count rows from zero instead of one.
    
Look at the entries for `age` and `cabin`, and you'll see the `NaN` symbol. That means we don't know the age and cabin number for this passenger. This may or may not be important; we'll have to consider if these numbers are important given the questions we're asking.  

## Understand the Data

Let's do some preliminary steps to understand more about our data. We learned earlier that each row in out dataset has 12 columns of data. We can do some simiple summarization to discover how these twelve columns look over the entire dataset.  
  
The function `describe` will give us some simple summary statistics. It's always a good place to start with any dataset.

In [7]:
titanic_df.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881135     0.498854     0.385027   
std       0.837836     0.486055    14.413500     1.041658     0.865560   
min       1.000000     0.000000     0.166700     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

We can see a few useful things in this description.
1) The first row of the table is labeled `count`, and it tells us how many entries are in each of the columns. In our `titanic_df` dataset, there are 1309 records...or passengers. All of the columns except 1 `age` show 1309 as the `count`.  
2) Notice that the `age ` column has 1046 listed for its `count`. That's due to missing data.
3) The second row of the table is labeled `mean`: Here's where we'll find the mean or average of each data column. For example, you can see the mean `age` is 29.881135. We'll might want to round that to 30 if we ever need to report it.
4) Remember from our data descriptions above that `survived` can take one of two values: 0 if the passenger didn't survive, 1 if they did. The `mean` for this column is 0.381971: 38% of the 1309 passengers in the table survived.
5) The passenger class, `pclass`, is another *categorical* variable: It can be first, second, or third (1, 2, or 3). `describe` computes a mean for this column, but it doesn't make sense because the values are categorical. The average here doesn't mean that someone sat in a 2.294882 class.

What can we do to get a better understanding of `pclass` and similar categorical variables? `pandas` has a function called `value_counts` that shows how many values are associated with each unique category in a column. For example, it might be useful to see how many passengers were in each passenger class.

In [8]:
titanic_df['pclass'].value_counts()

pclass
3    709
1    323
2    277
Name: count, dtype: int64

You can see that 323 passengers were in first class, 277 in second, and 709 sailed in third class. We can also find out how many passengers survived using `value_counts`. To do this, you need to specify a column in `titanic_df` that corresponds to the number of survivors.

In [22]:
# Use value_counts() to count the number of survivors.
survived_counts = titanic_df['survived'].value_counts()
print(survived_counts)

survived
0    809
1    500
Name: count, dtype: int64


You should see the numbers for those who survived and those who didn't.
Another useful trick is to "normalize" the numbers into percentages. We do that by setting the `normalize` parameter to True when we call `value_counts()`.

In [23]:
# Use value_counts(normalize=True) to get the percentages of survivors.
percentage_survived_counts = titanic_df['survived'].value_counts(normalize=True)
print(percentage_survived_counts)

survived
0    0.618029
1    0.381971
Name: proportion, dtype: float64


### <span style="color: red;">Question</span>
We started by asking three questions. Fill in the answers to these below:

In [44]:
from re import S
# 1. How many people were on the Titanic?
total_passengers = len(titanic_df)

# 2. How many survived the accident?
total_survivors = survived_counts.get(1)

# 3. How many fatalities were there?
total_fatalities = survived_counts.get(0)

# 4. What percentage of passengers survived?
percent_survivors = percentage_survived_counts.get(1).round(2) * 100

print(f"There were {total_passengers} passengers on the Titanic.")
print(f"{total_survivors} passengers survived the accident.")
print(f"{total_fatalities} passengers died in the accident.")
print(f"{percent_survivors}% of passengers survived the accident.")






There were 1309 passengers on the Titanic.
500 passengers survived the accident.
809 passengers died in the accident.
38.0% of passengers survived the accident.


### Basic Data Filtering
Another way to find the number of survivors is to access the `survived` column and filter out those rows that are equal to 1. Remember that the [data dictionary](#the-data-dictionary) specified that a "1" equals survived and a "0" equals didn't survive.  

So we can write a line of Python that returns only values where `survived == 1`.

In [45]:
survivors = titanic_df[titanic_df['survived'] == 1]

# use Python's len function to compute the length of the survivors data frame
total_number_of_survivors = len(survivors)

format(f"{total_number_of_survivors} people survived the disaster.")

'500 people survived the disaster.'

Similarly, we can ask how many passengers on the Titantic were male. For this, we want to look at the `sex` column and select only those rows where `sex == 'male'`. Here's the code:

In [46]:
male_passengers = titanic_df[titanic_df['sex'] == 'male']

# use len to count the male passengers
number_of_male_passengers = len(male_passengers)

format(f"{number_of_male_passengers} men were onboard the Titanic.")

'843 men were onboard the Titanic.'

You can combine any number of conditional statements in pandas to focus on specific data. For example, we could ask how many survivors were male? In which case, we want two conditional statements:
1. How many people survived (`survived == 1`), AND (&)
2. How many people were male (`sex == 'male'`)

In [47]:
male_survivors = titanic_df[(titanic_df['sex'] == 'male') & (titanic_df['survived'] == 1)]

# use len to count the number of male survivors
number_of_male_survivors = len(male_survivors)

print(f"{number_of_male_survivors} men survived the Titanic disaster.")

161 men survived the Titanic disaster.


It's your turn now. How many female passengers survived the accident? You'll want to combine the number of survivors (`survived`) AND the number of female passengers (`sex`).

In [49]:
female_survivors = titanic_df[(titanic_df['sex'] == 'female') & (titanic_df['survived'] == 1)]

# use Python's len to count the number of female survivors
number_of_female_survivors = len(female_survivors)

print(f"{number_of_female_survivors} women survived the Titanic disaster.")

339 women survived the Titanic disaster.


What about children? Let's compute the following:
1. How many children were on the Titanic? Let's consider "children" to be people 18 years old or under.
2. How many children survived the disaster?
3. What percentage of the women onboard survived? How about for men and childen?

In [55]:
# How many children were on the Titanic? Let's consider "children" to be people
# 18 years old or under.
children_data = titanic_df[titanic_df['age'] <= 18]
number_of_children = len(children_data)
print(f"There were {number_of_children} children onboard the Titanic.")

# How many children survived the disaster?
child_survivors = titanic_df[(titanic_df['age'] <= 18) & (titanic_df['survived'] == 1)]
number_of_child_survivors = len(child_survivors)
print(f"{number_of_child_survivors} of those children survived the accident.")

# What percentage of the women onboard survived?
female_passengers = titanic_df[titanic_df['sex'] == 'female']
number_of_female_passengers = len(female_passengers)
percent_female_survivors = number_of_female_survivors / number_of_female_passengers

# What percentage of the men onboard survived?
percent_male_survivors = (number_of_male_survivors / number_of_male_passengers)

# What percentage of the children onboard survived?
percent_child_survivors = (number_of_child_survivors / number_of_children)

print(f"{percent_female_survivors*100:.0f}% of the women onboard survived.")
print(f"{percent_male_survivors*100:.0f}% of the men onboard survived.")
print(f"{percent_child_survivors*100:.0f}% of the children onboard survived.")



There were 193 children onboard the Titanic.
95 of those children survived the accident.
73% of the women onboard survived.
19% of the men onboard survived.
49% of the children onboard survived.


As an optional exercise, you could compute the percentage of survivors who were women, men, or children.

## Cleaning the Data
Let's use the `.info()` function to see some properties of the DataFrame.

In [56]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


### The End...For Now

Now go back to Canvas to finish up our first look at the Titanic.
